### ISSUE TITLE
function signatures: support for providing CUE-based function implementation

### ISSUE THREAD

**Issue by myitcv** [cue-lang/cue#2183]
Creating this as a sub-issue of #2007.

My understanding thus far is that we will have some means of declaring function _signatures_:

```
foo: func(int): int
```

For then declaring "instances" (is this the right word?) in WASM, we would then have something like this:

```
// add5 adds 5 to a number
add5: wasm.Import(as: foo, wasm.bin)
```

Then `add5` could be used like this:

```
x: add5(4) // 9
```

(let's wave our hands at quite how we invoke `cmd/cue` or the Go API to make this happen, find the WASM etc).

It strikes me that we should also allow for people providing a function implementation in _CUE_. So instead of the WASM-based approach above (perhaps enabled/disabled via a build tag or similar), to offer a CUE-based alternative we would do:

```
add5: foo
add5: {
    #in: int
    #res: #in + 5
}
```

(ignore the fact I've done a pseudo-struct-style function implementation here in CUE, it could take any form).

The reality is that for simple functions it might well be that people do rough and ready implementations/approximations in CUE first, then upgrade at a later date. Having the ability to swap where the function is _implemented_ seems to be a key enabler to that end.

cc @4ad @rogpeppe

**Comment by mpvl** [comment]
I don't think this is urgent, or necessary at all, if the WASM bridge works out well.

Coding in CUE should generally be discouraged.

**Comment by mvdan** [comment]
For those following this thread, please see our proposal and experimental support for CUE functions: https://github.com/cue-lang/cue/discussions/4484
